tests/features/support/steps.js

```diff
@@ -51,38 +110,158 @@ When(/I click the (button|link) "([^"]+)"/, async (elemType, phrase) => {
       scope.page.waitForNavigation({waitUntil: 'domcontentloaded'})
-  ]);
+    ]);
   } else {
     if (process.env.DEBUG) {
-      await scope.page.screenshot({ path: './error.jpg', type: 'jpeg' });
+      await scope.page.screenshot({ path: './error.jpg', type: 'jpeg', fullPage: true });
     }
     throw `No ${elemType} with text ${phrase} exists.`;
   }
 });
 
 When('I click the defined text link {string}', async (phrase) => {
+  // Should we wait for navigation here?
   const [link] = await scope.page.$x(`//a[contains(text(), "${phrase}")]`);
   if (link) {
-    await link.click();
+    await link.click();  // TODO: change to `clickOrTap`
   } else {
     if (process.env.DEBUG) {
-      await scope.page.screenshot({ path: './error.jpg', type: 'jpeg' });
+      await scope.page.screenshot({ path: './error.jpg', type: 'jpeg', fullPage: true });
     }
     throw `No link with text ${phrase} exists.`;
   }
 });
 
+Then('I should see the link {string}', async (linkText) => {
+  let [link] = await scope.page.$x(`//a[contains(text(), "${linkText}")]`);
+  expect(link).to.exist;
+});
+
+// TODO: Research how da handles invalid symbols
+Then('the question id should be {string}', async (question_class) => {
+  /* Looks for the question id as it's written in the .yml.
+  *  
+  *  WARNING: Does not handle actual html class attribute name on page
+  */
+  question_class = question_class.replace(/\s/g, '-');
+  question_class = 'question-' + question_class;
+  const element = await scope.page.waitFor('body.' + question_class);
+  expect(element).to.exist;
+});
+
+Then('an element should have the id {string}', async (id) => {
+  const element = await scope.page.waitFor('#' + id);
+  expect(element).to.exist;
+});
+
 Then('I should see the phrase {string}', async (phrase) => {
   const bodyText = await scope.page.$eval('body', elem => elem.innerText);
   expect(bodyText).to.contain(phrase);
 });
 
+Then(/the link "([^"]+)" should lead to "([^"]+)"/, async (linkText, expected_url) => {
+  let [link] = await scope.page.$x(`//a[contains(text(), "${linkText}")]`);
+  
+  let prop_obj = await link.getProperty('href');
+  let actual_url = await prop_obj.jsonValue();
+  expect( actual_url ).to.equal( expected_url );
+});
+
+Then(/the link "([^"]+)" should open a working page/, async (linkText) => {
+  let [link] = await scope.page.$x(`//a[contains(text(), "${linkText}")]`);
+  let prop_obj = await link.getProperty('href');
+  let actual_url = await prop_obj.jsonValue();
+  
+  let linkPage = await scope.browser.newPage();
+  let response = await linkPage.goto(actual_url, {waitUntil: 'domcontentloaded'});
+  expect( response.ok() ).to.be.true;
+  linkPage.close()
+});
+
+Then(/the link "([^"]+)" should open in (a new window|the same window)/, async (linkText, which_window) => {
+  let [link] = await scope.page.$x(`//a[contains(text(), "${linkText}")]`);
+
+  let prop_obj = await link.getProperty('target');
+  let target = await await prop_obj.jsonValue();
```

Comment:
Should this be clicking on the link? Right now, I think it's just checking whether the link opens in a new window or not.

Reply:
There was some complication involved in clicking on the link. I don't remember what. Maybe it was that some buttons, like the escape button, won't open a new window - just change the current one, so I didn't want links in general to actually be clicked on. Maybe it was something else? Maybe I had trouble detecting a new window opening with puppeteer?

To sum up, it is deliberately not clicking on the link, though it might be worth discussing alternatives. Perhaps for a future issue? (I only mention making issues because if this is pulled in, other development can happen at the same time as this is being improved.)

Reply:
I realize you've approved, but I want to be clear that I'm not dismissing these. They've all been good to bring up.